### PR TITLE
ThemeStore: Adding tests (and only tests)

### PR DESCRIPTION
**Pull request by tonyr59h**
This PR comprises most of the changes from #582. It adds various unit tests and release tests for Jetpack and WP.com actions.

To test:
* Update your `tests.properties`
* Run the tests

cc @oguzkocer

### CODE REVIEW

File: example/src/test/java/org/wordpress/android/fluxc/theme/ThemeStoreUnitTest.java [hunk 1]
```diff
@@ -38,7 +40,55 @@ public void setUp() {
...
+        assertNull(mThemeStore.getActiveThemeForSite(site));
+
+        final ThemeModel firstTheme = generateTestTheme(site.getId(), "first-active", "First Active");
+        final ThemeModel secondTheme = generateTestTheme(site.getId(), "second-active", "Second Active");
+        firstTheme.setActive(true);
+        secondTheme.setActive(true);
+
+        // set first theme active and verify
+        mThemeStore.setActiveThemeForSite(site, firstTheme);
+        ThemeModel firstStoreTheme = mThemeStore.getActiveThemeForSite(site);
```
Comment (oguzkocer) [inline example/src/test/java/org/wordpress/android/fluxc/theme/ThemeStoreUnitTest.java]: I don't think this is the correct approach to this test as `ThemeStore` will only return the first active theme. We should be checking that `firstTheme` is the **only** active theme and the way to do that is to use `ThemeSqlUtils.getActiveThemeForSite(site)` and make sure the list contains only 1 theme and that's equal to `firstTheme`.

We should do the same thing for the `secondTheme` as well.

File: example/src/test/java/org/wordpress/android/fluxc/theme/ThemeStoreUnitTest.java [hunk 2]
```diff
@@ -38,7 +40,55 @@ public void setUp() {
...
+
+    @Test
+    public void testInsertOrUpdateTheme() {
+        final String testThemeId = "fluxc-ftw";
+        final String testThemeName = "FluxC FTW";
+        final String testUpdatedName = testThemeName + " v2";
+        final ThemeModel insertTheme = generateTestTheme(12345, testThemeId, testThemeName);
+
+        // verify theme doesn't already exist
+        assertEquals(null, mThemeStore.getThemeByThemeId(testThemeId, false));
```
Comment (oguzkocer) [inline example/src/test/java/org/wordpress/android/fluxc/theme/ThemeStoreUnitTest.java]: While reviewing this test, I am realizing that maybe it'd have been better to have 2 methods: `getInstalledThemeByThemeId` and `getWpComThemeByThemeId` and have them call the `getThemeByThemeId` private method. I think it might be better for the consumers of the library. What do you think?

Comment (oguzkocer) [reply example/src/test/java/org/wordpress/android/fluxc/theme/ThemeStoreUnitTest.java]: Also, we can use `assertNull` here instead :)

File: example/src/test/java/org/wordpress/android/fluxc/theme/ThemeStoreUnitTest.java [hunk 3]
```diff
@@ -38,7 +40,55 @@ public void setUp() {
...
+
+        // insert new theme and verify it exists
+        ThemeSqlUtils.insertOrUpdateThemeForSite(insertTheme);
+        assertNotNull(mThemeStore.getThemeByThemeId(testThemeId, false));
```
Comment (oguzkocer) [inline example/src/test/java/org/wordpress/android/fluxc/theme/ThemeStoreUnitTest.java]: We should extract `mThemeStore.getThemeByThemeId(testThemeId, false)` into a variable so we don't have to call it again in the next line.

File: example/src/test/java/org/wordpress/android/fluxc/theme/ThemeStoreUnitTest.java [hunk 4]
```diff
@@ -38,7 +40,55 @@ public void setUp() {
...
+        assertEquals(testThemeName, mThemeStore.getThemeByThemeId(testThemeId, false).getName());
+
+        // update the theme and verify the updated attributes
+        insertTheme.setName(testUpdatedName);
+        ThemeSqlUtils.insertOrUpdateThemeForSite(insertTheme);
+        assertNotNull(mThemeStore.getThemeByThemeId(testThemeId, false));
```
Comment (oguzkocer) [inline example/src/test/java/org/wordpress/android/fluxc/theme/ThemeStoreUnitTest.java]: Similar to above, let's extract `mThemeStore.getThemeByThemeId(testThemeId, false)` into a variable.

File: example/tests.properties-example [hunk 5]
```diff
@@ -12,6 +12,10 @@ TEST_WPCOM_IMAGE_ID_TO_CHANGE = FIXME
 TEST_WPCOM_USERNAME_2FA = FIXME
 TEST_WPCOM_PASSWORD_2FA = FIXME
 
+# WP.com - account with purchased themes
+TEST_WPCOM_USERNAME_PURCHASED_THEMES = FIXME
```
Comment (oguzkocer) [inline example/tests.properties-example]: These properties doesn't seem to be used either.

File: example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestWPCom.java [hunk 6]
```diff
@@ -27,13 +47,108 @@ protected void setUp() throws Exception {
...
 
-    public void testFetchWpComThemes() throws InterruptedException {
+    public void testActivateTheme() throws InterruptedException {
+        // get all themes available
         mCountDownLatch = new CountDownLatch(1);
         mNextEvent = TestEvents.FETCHED_THEMES;
         mDispatcher.dispatch(ThemeActionBuilder.newFetchWpComThemesAction());
         assertTrue(mCountDownLatch.await(TestUtils.DEFAULT_TIMEOUT_MS, TimeUnit.MILLISECONDS));
+        List<ThemeModel> themes = mThemeStore.getWpComThemes();
+        assertTrue(themes.size() > 1);
```
Comment (oguzkocer) [inline example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestWPCom.java]: Shouldn't we check for more than 2 themes here?

Comment (oguzkocer) [reply example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestWPCom.java]: Actually, since we need to get `getNewNonPremiumTheme` that doesn't even cover it. Maybe we can remove the whole thing and the test will just fail if the user doesn't have enough themes?

File: example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestWPCom.java [hunk 7]
```diff
@@ -27,13 +47,108 @@ protected void setUp() throws Exception {
...
+
+    public void testFetchWPComThemes() throws InterruptedException {
+        // no need to sign into account, this is a test for an endpoint that does not require authentication
+        mCountDownLatch = new CountDownLatch(1);
+        mNextEvent = TestEvents.FETCHED_THEMES;
+        mDispatcher.dispatch(ThemeActionBuilder.newFetchWpComThemesAction());
+
+        // verify response received and WP themes list is not empty
+        assertTrue(mCountDownLatch.await(TestUtils.DEFAULT_TIMEOUT_MS, TimeUnit.MILLISECONDS));
+        assertTrue(mThemeStore.getWpComThemes().size() > 0);
```
Comment (oguzkocer) [inline example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestWPCom.java]: Maybe we should check that `mThemeStore.getWpComThemes()` is empty before doing the check to make sure the test actually did something.

File: example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestWPCom.java [hunk 8]
```diff
@@ -27,13 +47,108 @@ protected void setUp() throws Exception {
...
+    }
+
+    @SuppressWarnings("unused")
+    @Subscribe
+    public void onThemeActivated(ThemeStore.OnThemeActivated event) {
+        if (event.isError()) {
+            throw new AssertionError("Unexpected error occurred with type: " + event.error.type);
+        }
+        assertTrue(mNextEvent == TestEvents.ACTIVATED_THEME);
+        mActivatedTheme = event.theme;
```
Comment (oguzkocer) [inline example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestWPCom.java]: I've already left some comments about `onThemeActivated` in the Jetpack tests, so I'll skip them here. However, I wanted to ask if this is actually an activation, should we be setting `mCurrentTheme` here as well? Obviously it won't work right now, but if we change that behaviour, I'd expect the currentTheme to be the activated theme as well.

File: example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestWPCom.java [hunk 9]
```diff
@@ -42,10 +157,107 @@ public void onThemesChanged(ThemeStore.OnThemesChanged event) {
...
+    @Subscribe
+    public void onSiteRemoved(SiteStore.OnSiteRemoved event) {
+        if (event.isError()) {
+            throw new AssertionError("Unexpected error occurred with type: " + event.error.type);
+        }
+        assertEquals(TestEvents.SITE_REMOVED, mNextEvent);
+        mCountDownLatch.countDown();
+    }
+
+    private void authenticateWPComAndFetchSites(String username, String password) throws InterruptedException {
```
Comment (oguzkocer) [inline example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestWPCom.java]: I think we can avoid this whole thing if we based this class from `ReleaseStack_WPComBase`. However, let's do this in a separate PR as it'd be much easier to review there.

File: example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestJetpack.java [hunk 10]
```diff
@@ -0,0 +1,374 @@
...
+        SiteModel jetpackSite = getJetpackSite();
+        assertNotNull(jetpackSite);
+
+        // get installed themes
+        mCountDownLatch = new CountDownLatch(1);
+        mNextEvent = TestEvents.FETCHED_INSTALLED_THEMES;
+        mDispatcher.dispatch(ThemeActionBuilder.newFetchInstalledThemesAction(jetpackSite));
+        assertTrue(mCountDownLatch.await(TestUtils.DEFAULT_TIMEOUT_MS, TimeUnit.MILLISECONDS));
+        List<ThemeModel> themes = mThemeStore.getThemesForSite(jetpackSite);
+        assertTrue(themes.size() > 1);
```
[the remark this answers is not in the document]
Comment (tonyr59h) [reply example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestJetpack.java]: Yep, that's `> 1` not `>= 1` 😉

File: example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestJetpack.java [hunk 11]
```diff
@@ -0,0 +1,374 @@
...
+        assertTrue(mThemeStore.getThemesForSite(jetpackSite).isEmpty());
+
+        // get installed themes
+        mCountDownLatch = new CountDownLatch(1);
+        mNextEvent = TestEvents.FETCHED_INSTALLED_THEMES;
+        mDispatcher.dispatch(ThemeActionBuilder.newFetchInstalledThemesAction(jetpackSite));
+        assertTrue(mCountDownLatch.await(TestUtils.DEFAULT_TIMEOUT_MS, TimeUnit.MILLISECONDS));
+
+        List<ThemeModel> themes = mThemeStore.getThemesForSite(jetpackSite);
+        assertFalse(themes.isEmpty());
```
[the remark this answers is not in the document]
Comment (tonyr59h) [reply example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestJetpack.java]: This check is to make sure the installed themes were successfully fetched/stored. Thanks for asking because there's a potential bug here if `Edin` is the only installed theme, fix incoming in another PR.

File: example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestJetpack.java [hunk 12]
```diff
@@ -0,0 +1,374 @@
...
+        init();
+        // Reset expected test event
+        mNextEvent = TestEvents.NONE;
+        mCurrentTheme = null;
+        mActivatedTheme = null;
+    }
+
+    public void testFetchInstalledThemes() throws InterruptedException {
+        // sign into a WP.com account with a Jetpack site
+        authenticateWPComAndFetchSites(BuildConfig.TEST_WPCOM_USERNAME_SINGLE_JETPACK_ONLY,
```
[the remark this answers is not in the document]
Comment (tonyr59h) [reply example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestJetpack.java]: c9d06adc823365f293460d10ba9647a9dfcca76c

File: example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestJetpack.java [hunk 13]
```diff
@@ -0,0 +1,374 @@
...
+        assertTrue(mCountDownLatch.await(TestUtils.DEFAULT_TIMEOUT_MS, TimeUnit.MILLISECONDS));
+        assertNotNull(mCurrentTheme);
+
+        // activate a different theme
+        ThemeModel themeToActivate = mCurrentTheme.getThemeId().equals(themes.get(0).getThemeId())
+                ? themes.get(1) : themes.get(0);
+        assertNotNull(themeToActivate);
+        mCountDownLatch = new CountDownLatch(1);
+        mNextEvent = TestEvents.ACTIVATED_THEME;
+        ThemeStore.ActivateThemePayload payload = new ThemeStore.ActivateThemePayload(jetpackSite, themeToActivate);
```
[the remark this answers is not in the document]
Comment (tonyr59h) [reply example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestJetpack.java]: e2922f442f7dd6b99a0d6a7e33534368c5f9e0ff

File: example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestJetpack.java [hunk 14]
```diff
@@ -0,0 +1,374 @@
...
+        mDispatcher.dispatch(ThemeActionBuilder.newActivateThemeAction(payload));
+        assertTrue(mCountDownLatch.await(TestUtils.DEFAULT_TIMEOUT_MS, TimeUnit.MILLISECONDS));
+        assertNotNull(mActivatedTheme);
```
[the remark this answers is not in the document]
Comment (tonyr59h) [reply example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestJetpack.java]: 313d97e4e8a50512dac84e3654e7e70ed35d92ab

File: example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestJetpack.java [hunk 15]
```diff
@@ -0,0 +1,374 @@
...
+        final String themeId = "edin-wpcom";
+        final ThemeModel themeToInstall = new ThemeModel();
+        themeToInstall.setName("Edin");
+        themeToInstall.setThemeId(themeId);
+
+        authenticateWPComAndFetchSites(BuildConfig.TEST_WPCOM_USERNAME_SINGLE_JETPACK_ONLY,
+                BuildConfig.TEST_WPCOM_PASSWORD_SINGLE_JETPACK_ONLY);
+        SiteModel jetpackSite = getJetpackSite();
+        assertNotNull(jetpackSite);
+        assertTrue(mThemeStore.getThemesForSite(jetpackSite).isEmpty());
```
[the remark this answers is not in the document]
Comment (tonyr59h) [reply example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestJetpack.java]: Unneeded, removed in d1e03b688dc58136b9ea126a61f81fc7aef84d28

File: example/src/test/java/org/wordpress/android/fluxc/theme/ThemeStoreUnitTest.java (hunk 3 above)
Comment (tonyr59h) [reply example/src/test/java/org/wordpress/android/fluxc/theme/ThemeStoreUnitTest.java]: 469600d4e6e3518ea2e298a2e4cc938d3bdfa862

File: example/src/test/java/org/wordpress/android/fluxc/theme/ThemeStoreUnitTest.java (hunk 4 above)
Comment (tonyr59h) [reply example/src/test/java/org/wordpress/android/fluxc/theme/ThemeStoreUnitTest.java]: 469600d4e6e3518ea2e298a2e4cc938d3bdfa862

File: example/tests.properties-example (hunk 5 above)
Comment (tonyr59h) [reply example/tests.properties-example]: Oops, these were added before I decided not to implement the `*_PURCHASED_THEMES` actions. Removed in e5a2071f3ccb08830e3da592144aca8a96d437e8

File: example/tests.properties-example [hunk 16]
```diff
@@ -182,6 +186,12 @@ TEST_WPORG_PASSWORD_JETPACK_SUBFOLDER = FIXME
 TEST_WPORG_URL_JETPACK_SUBFOLDER = FIXME
 TEST_WPORG_URL_JETPACK_SUBFOLDER_ENDPOINT = FIXME
 
+# Self hosted - Jetpack site with two themes installed
+TEST_WPCOM_USERNAME_JETPACK_TWO_THEMES = FIXME
```
[the remark this answers is not in the document]
Comment (tonyr59h) [reply example/tests.properties-example]: I opted to hard code instead of using `TEST_THEME1/2`.

I actually thought these credentials were the ones I was using in the tests 😁 

Removed in e5a2071f3ccb08830e3da592144aca8a96d437e8

File: example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestJetpack.java [hunk 17]
```diff
@@ -0,0 +1,374 @@
...
+
+        // get installed themes
+        mCountDownLatch = new CountDownLatch(1);
+        mNextEvent = TestEvents.FETCHED_INSTALLED_THEMES;
+        mDispatcher.dispatch(ThemeActionBuilder.newFetchInstalledThemesAction(jetpackSite));
+        assertTrue(mCountDownLatch.await(TestUtils.DEFAULT_TIMEOUT_MS, TimeUnit.MILLISECONDS));
+
+        List<ThemeModel> themes = mThemeStore.getThemesForSite(jetpackSite);
+        assertFalse(themes.isEmpty());
```
[the remark this answers is not in the document]
Comment (tonyr59h) [reply example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestJetpack.java]: Need to make sure we have the installed themes so we don't attempt to install a duplicate.

File: example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestWPCom.java (hunk 8 above)
Comment (tonyr59h) [reply example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestWPCom.java]: Makes sense to me, added in 173ad6c47dc740fa5cd489287642235f9ecea167

File: example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestJetpack.java [hunk 18]
```diff
@@ -0,0 +1,374 @@
...
+        mCountDownLatch = new CountDownLatch(1);
+        mNextEvent = TestEvents.FETCHED_CURRENT_THEME;
+        mDispatcher.dispatch(ThemeActionBuilder.newFetchCurrentThemeAction(jetpackSite));
+        assertTrue(mCountDownLatch.await(TestUtils.DEFAULT_TIMEOUT_MS, TimeUnit.MILLISECONDS));
+
+        // if Edin is active update site's active theme to something else
+        if (themeId.equals(mCurrentTheme.getThemeId())) {
+            mCountDownLatch = new CountDownLatch(1);
+            mNextEvent = TestEvents.ACTIVATED_THEME;
+            ThemeModel toActivate = getOtherTheme(themes, themeId);
```
[the remark this answers is not in the document]
Comment (tonyr59h) [reply example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestJetpack.java]: 3744454c7540309402f01b7fc407f5af2557548c

File: example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestWPCom.java (hunk 6 above)
Comment (tonyr59h) [reply example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestWPCom.java]: > Maybe we can remove the whole thing and the test will just fail if the user doesn't have enough themes?

Good point, this is redundant. Dropped in f1e7a3c06dd0fb03c54704b664e62f632578508e

File: example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestJetpack.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (oguzkocer) [reply example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_ThemeTestJetpack.java]: 🤦‍♂️
